src/jdk.jdwp.agent/share/native/libjdwp/log_messages.c

```diff
@@ -78,9 +78,11 @@ get_time_stamp(char *tbuf, size_t ltbuf)
                 "%d.%m.%Y %T", localtime(&t));
     (void)strftime(timestamp_timezone, TZ_SIZE,
                 "%Z", localtime(&t));
-    (void)snprintf(tbuf, ltbuf,
-                   "%s.%.3d %s", timestamp_date_time,
-                   (int)(millisecs), timestamp_timezone);
+    // Truncate milliseconds in buffer large enough to hold the
+    // value which is always < 1000
```

Comment:
I initially missed the %d to %s change. Can we augment the comment to say:

// value which is always < 1000 (and so a maximum of 3 digits for "%.3s")

Also I'm not clear whether this will result in extra spaces when the ms value is < 100 ?

Reply:
Right, `snprintf(tmp, sizeof(tmp), "%d", millisecs);` needs to be 
`snprintf(tmp, sizeof(tmp), "%.3d", millisecs);` to pad it out correctly.

Reply:
timestamp 05.11.2020 19:54:13.100 EST

This is what I get (in a small test program) for 100 ms with the existing code.  Is this not right?

Reply:
char tmp[10 + 1];
    snprintf(tmp, sizeof(tmp), "%.3d", millisecs);
    snprintf(tbuf, ltbuf, "%s.%s %s", timestamp_date_time, tmp, timestamp_timezone);
This also gets the same thing.

Reply:
The concern is when it is less than 100ms.

Reply:
unsigned millisecs[] = { 2, 20, 200, 1000 };
  get_time_stamp(millisecs[i], buf, sizeof(buf));

gets:

timestamp 06.11.2020 06:56:08.002 EST
timestamp 06.11.2020 06:56:08.020 EST
timestamp 06.11.2020 06:56:08.200 EST
timestamp 06.11.2020 06:56:08.1000 EST

with

    char tmp[10 + 1];
    snprintf(tmp, sizeof(tmp), "%.3d", millisecs);
    snprintf(tbuf, ltbuf, "%s.%s %s", timestamp_date_time, tmp, timestamp_timezone);

Is this what you want?

Reply:
I think you need .3 in both places, otherwise I expect the warning will still be there? (We don't need to worry about values of millisecs larger than 999):

```char tmp[11 + 1];
snprintf(tmp, sizeof(tmp), "%.3d", millisecs);
snprintf(tbuf, ltbuf, "%s.%.3s %s", timestamp_date_time, tmp, timestamp_timezone);```

Reply:
You're right.